sensors/cmd/start.go

```diff
@@ -70,20 +45,194 @@ func Start() {
+	return &SensorContextManager{
+		C:               make(chan func() error),
+		logger:          logger,
+		kubeClient:      kubeClient,
+		dynamicClient:   dynamicClient,
+		eventBusConfig:  eventBusConfig,
+		eventBusSubject: eventBusSubject,
+		hostname:        hostname,
+	}
+}
+
+func (scm *SensorContextManager) Start(ctx context.Context) error {
+	// SENSOR_PATH env variable is used to indicate if this sensor definition
+	// has opted in to live reload
+	path, liveReload := os.LookupEnv("SENSOR_PATH")
+
+	switch liveReload {
+	case true:
+		return scm.startLiveReload(ctx, path)
+	case false:
+		return scm.startDefault(ctx)
+	}
+
+	return nil
+}
+
+func (scm *SensorContextManager) Close() {
+	close(scm.C)
+}
+
+func (scm *SensorContextManager) startDefault(ctx context.Context) error {
+	sensor, err := scm.readSensorFromEnv()
+	if err != nil {
+		scm.logger.Fatalw("failed to read sensor from env variable", zap.Error(err))
 	}
 
-	ctx := logging.WithLogger(signals.SetupSignalHandler(), logger)
 	m := metrics.NewMetrics(sensor.Namespace)
 	go m.Run(ctx, fmt.Sprintf(":%d", common.SensorMetricsPort))
 
-	logger.Infow("starting sensor server", "version", argoevents.GetVersion())
-	sensorExecutionCtx := sensors.NewSensorContext(kubeClient, dynamicClient, sensor, busConfig, ebSubject, hostname, m)
-	if err := sensorExecutionCtx.Start(ctx); err != nil {
-		logger.Fatalw("failed to listen to events", zap.Error(err))
+	sensorCtx := sensors.NewSensorContext(scm.kubeClient, scm.dynamicClient, sensor, scm.eventBusConfig, scm.eventBusSubject, scm.hostname, m)
+	scm.C <- func() error { return sensorCtx.Start(ctx) }
+
+	return nil
+}
+
+func (scm *SensorContextManager) startLiveReload(ctx context.Context, path string) error {
+	// watch for sensor file changes
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	err = watcher.Add(path)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer watcher.Close()
+		var m *metrics.Metrics
+
+		for {
+			sensor, err := scm.readSensorFromFile(path)
+			if err != nil {
+				scm.logger.Fatalw("failed to read sensor from file", zap.Error(err))
+			}
+
+			// logger
+			scm.logger = scm.logger.With("sensorName", sensor.Name)
+			for name, value := range sensor.Spec.LoggingFields {
+				scm.logger.With(name, value)
+			}
+
+			// start metrics once
+			if m == nil {
+				m = metrics.NewMetrics(sensor.Namespace)
+				go m.Run(ctx, fmt.Sprintf(":%d", common.SensorMetricsPort))
+			}
+
+			// create context
+			subCtx, cancel := context.WithCancel(ctx)
+			sensorCtx := sensors.NewSensorContext(scm.kubeClient, scm.dynamicClient, sensor, scm.eventBusConfig, scm.eventBusSubject, scm.hostname, m)
+			scm.C <- func() error { return sensorCtx.Start(subCtx) }
+
+			for {
+				select {
+				case event := <-watcher.Events:
+					if event.Op == fsnotify.Write || event.Op == fsnotify.Remove {
+						cancel()
+
+						// _ = watcher.Remove(path)
```

Comment:
Does the code need lines 178-181 to remove the path from the watcher and add it back? Line 178 to remove is already commented out

Reply:
I played around with, we don't need to remove so I have removed that line entirely. We do need to add it back because the kubelet does some sort of atomic file swap for volumne mounts where the file is effectively deleted and re-created. We have to call add again so that we watch the _new_ file.